### PR TITLE
Add support for gzip

### LINKED ISSUE #616: Support gzip encoding on eth1 http json rpc
## Description

Presently we do not handle gzip encoding when connecting to an eth1 peer. Gzip is commonly used by public Ethereum nodes because it significantly reduces the size of resoonses.

## Detail

I provide details in the comments in the code used to communicate with the eth1 node via HTTP:

https://github.com/sigp/lighthouse/blob/24e941d175a7455209c140157bf8d5d588c87ba1/beacon_node/eth1/src/http.rs#L344-L364

Sometime after my first attempt at this I noticed that `reqwest` (the http client library we use) has a [`gzip method`](https://docs.rs/reqwest/0.9.22/reqwest/struct.ClientBuilder.html#method.gzip). I suspect this will solve all our problems.

## Steps to resolve

- Fix gzip encoding.
- Prove that it works, by either;
   - Adding unit tests.
   - If it's too hard to add a test (because you need to build a gzip http server), then provide some details on how one could prove it works (e.g., run some command and it see it communicate with some public server).

### PR DESCRIPTION
**Pull request by pscott**
## Issue Addressed

Closes #616 

## Proposed Changes

- Add support for gzip
- Fix clippy warnings
- Use start and end instead of range in `get_deposits` parameters

## Additional Info

Testing this was quite a challenge: the [doc](https://docs.rs/reqwest/0.9.22/reqwest/struct.ClientBuilder.html#method.gzip) clearly states that:
1. gzip is enabled by default
2. gzip will add the appropriate headers `accept-encoding: gzip` if no other accept-encoding header is present and range is not set.
3. When receiving a response, it will automatically REMOVE the content-encoding header and deflate the body.

For testing purposes, I used a different endpoint (https://goerli.prylabs.net), added the `gzip(false)` method and set the headers to `accept-encoding: gzip`. The response we were receiving were indeed gzip encoded.
I then used the code as presented in this PR, and ran the tests in the http module. They were the same whether I was using this endpoint or another endpoint; I also ran the tests with gzip set to false, to make sure the responses we got were not gzip encoded. I got the same errors (deposit count not increasing, etc..) whether I was using gzip or json (which is expected since the deposit contract doesn't exist on the blockchain yet (?)).

Side note: while fixing clippy warnings, I ended up reworking the `get_deposits` function to use some `start, end` parameters rather than a range.

### CODE REVIEW

File: beacon_node/eth1/src/deposit_cache.rs [hunk 1]
```diff
@@ -143,24 +142,25 @@ impl DepositCache {
     ///
     /// ## Errors
     ///
-    /// - If `deposit_count` is larger than `range.end`.
+    /// - If `deposit_count` is larger than `end`.
     /// - There are not sufficient deposits in the tree to generate the proof.
     pub fn get_deposits(
         &self,
-        range: Range<u64>,
+        start: u64,
```
Comment (paulhauner) [inline beacon_node/eth1/src/deposit_cache.rs]: Hey, out of interest why did you change this? :)

Comment (pscott) [reply beacon_node/eth1/src/deposit_cache.rs]: So I was getting some clippy warnings re: this line 
https://github.com/sigp/lighthouse/blob/4b4bc6247da72a85154b3f7984f7ae917153db73/beacon_node/eth1/src/deposit_cache.rs#L363
saying "hey why are you not using `0..=n` instead of having `0..n + 1`. Classic range clippy warning.
However once I changed it, well it would not be a valid argument parameter since `get_deposit` was expecting a `Range`, not an `InclusiveRange`. So I thought about using the trait `RangeBounds`, and I did, and when I started refactoring `get_deposits` I thought "wait a minute, there's only one purpose to the range provided: know the start and the end of it".

So I thought it would be both more ergonomical and efficient (not acually generating a range before passing it to get_deposits) to use `start` and `end` u64 to do just that! :)

**Comment by paulhauner** [comment]
This seems like it's failing the CI in your repo:

https://github.com/pscott/lighthouse/runs/323645468

**Comment by pscott** [comment]
@paulhauner Fixed this silly mistake, however CI failed for [some obscure reasons](https://github.com/pscott/lighthouse/runs/342859729)?

**Comment by michaelsproul** [comment]
Looks like a libp2p dependency issue, likely fixed by merging in `master` (or rebasing on top).

**Comment by paulhauner** [comment]
Hey, thanks for fixing this! Can I please be a pain and ask you to merge `sim-ci` onto this?

Then we can get the simulator to ensure that the beacon node and val client can still talk to each other.

**Comment by pscott** [comment]
Done!
